intake/source/base.py

```diff
@@ -53,7 +52,133 @@ class NoEntry(AttributeError):
     pass
 
 
-class DataSource(DictSerialiseMixin):
+class CacheMixin:
+    _cache = None
+
+    @property
+    def cache_dirs(self):
+        return [c._cache_dir for c in self.cache]
+
+    def set_cache_dir(self, cache_dir):
+        for c in self.cache:
+            c._cache_dir = make_path_posix(cache_dir)
+
+    @property
+    def cache(self):
+        if self._cache is None:
+            self._cache = make_caches(self.name, self.metadata.get('cache'),
+                                      catdir=self.metadata.get('catalog_dir',
+                                                             None),
+                                      storage_options=self.storage_options)
+        return self._cache
+
+    @cache.setter
+    def cache(self, csh):
+        self._cache = csh
+
+    def _get_cache(self, urlpath):
```

Comment:
Can we also make this public?  It might make also make sense to put a no-op version of this on `DataSourceBase` that always returns `[urlpath]` so that consumers don't have to care if there is an actual cache or not.

I am also a bit confused about how the cache actually gets used?  It looks like this is the method that is called in the concrete implementations.  This method then in turn calls `c.load` which defers to the Cache objects and either fetch and store the data, returning a new url (or list of urls?) that point to the cached data.  We do this for all of the caches that are set up, but everyplace this is called we do `self._get_cache(urlpath)[0]` which only looks at the first entry.  I think this means we can spend time / bits caching things we are never going to look at?

Reply:
On a bit more consideration, would it be better to keep this private, but mix it in at the concrete source level?

Reply:
> would it be better to keep this private, but mix it in at the concrete source level?

In the long run yes, and I want to firmly deprecate this in favour of handling at the fsspec layer anyway. For this PR, I didn't want the actual functionality to move around too much.

Reply:
Let's leave this as future work. For now, I think we'll update the docs, sayin that new drivers can derive from DataSourceBase any any mixins they like, or DataSource to get them all.